cloudify/workflows/local.py

```diff
@@ -424,22 +416,31 @@ def update_node_instance(self,
                 instance['state'] = state
             self._store_instance(instance)
 
-    def _get_node_instance(self, node_instance_id):
+    def _get_node_instance(self, node_instance_id, evaluate_functions=True):
         instance = self._load_instance(node_instance_id)
         if instance is None:
-            raise RuntimeError('Instance {0} does not exist'
-                               .format(node_instance_id))
+            raise KeyError('Instance {0} does not exist'
+                           .format(node_instance_id))
+        if evaluate_functions:
+            dsl_functions.evaluate_node_instance_functions(
+                instance, self)
         return instance
 
-    def get_node(self, node_id):
+    def get_node(self, node_id, evaluate_functions=True):
         node = self._nodes.get(node_id)
         if node is None:
-            raise RuntimeError('Node {0} does not exist'
-                               .format(node_id))
+            raise KeyError('Node {0} does not exist'.format(node_id))
+        if evaluate_functions:
+            dsl_functions.evaluate_node_functions(node, self)
         return copy.deepcopy(node)
 
-    def get_nodes(self):
-        return copy.deepcopy(self._nodes.values())
+    def get_nodes(self, evaluate_functions=True):
+        nodes = copy.deepcopy(self._nodes.values())
```

Comment:
Maybe only do the deepcopy after deciding whether we can short circuit due to evaluate_functions=False?

Reply:
the nodes are handed out to plugins, which can mutate them, so they have to be deepcopied